src/v/utils/external_process.h

```diff
@@ -0,0 +1,262 @@
+    external_process& operator=(const external_process&) = delete;
+    external_process(external_process&&) noexcept;
+    external_process& operator=(external_process&&) noexcept = delete;
+
+    ~external_process() {
+        vassert(_completed, "Process has not yet completed");
+        vassert(
+          !_stdout_consumer.has_value(), "stdout consumer not cleaned up");
+        vassert(
+          !_stderr_consumer.has_value(), "stderr consumer not cleaned up");
+    }
+
+    /**
+     * @return true If process is still running, false otherwise
+     */
+    bool is_running() const { return !_completed; }
+
+    /**
+     * @return const std::filesystem::path& Path to the executable being
+     * executed
+     */
+    const std::filesystem::path& path() const noexcept { return _path; }
+    /**
+     * @return const ss::experimental::spawn_parameters& The spawn parameters
+     */
+    const ss::experimental::spawn_parameters& parameters() const noexcept {
+        return _params;
+    }
+
+    /**
+     * @brief Set the stdout consumer object
+     *
+     * @tparam Consumer The type of consumer, must satisfiy the requirements of
+     * ss::InputStreamConsumer
+     * @param consumer Instance of the consumer that will receive the output of
+     * stdout
+     * @note Will assert if called >1
+     */
+    template<typename Consumer>
+    requires ss::InputStreamConsumer<Consumer, char>
+    void set_stdout_consumer(Consumer consumer) {
+        vassert(!_stdout_consumer.has_value(), "Already set stdout consumer");
+        auto cout = _process.cout();
+        _stdout_consumer.emplace(ss::do_with(
+          std::move(cout),
+          [consumer = std::move(consumer)](
+            ss::input_stream<char>& cout) mutable {
+              return cout.consume(std::move(consumer));
+          }));
+    }
+
+    /**
+     * @brief Set the stderr consumer object
+     *
+     * @tparam Consumer The type of consumer, must satisfiy the requirements of
+     * ss::InputStreamConsumer
+     * @param consumer Instance of the consumer that will receive the output of
+     * stderr
+     * @note Will assert if called >1
+     */
+    template<typename Consumer>
+    requires ss::InputStreamConsumer<Consumer, char>
+    void set_stderr_consumer(Consumer consumer) {
+        vassert(!_stderr_consumer.has_value(), "Already set stdout consumer");
+        auto cerr = _process.cerr();
+        _stderr_consumer.emplace(ss::do_with(
+          std::move(cerr),
+          [consumer = std::move(consumer)](
+            ss::input_stream<char>& cerr) mutable {
+              return cerr.consume(std::move(consumer));
+          }));
+    }
+
+    /**
+     * @brief Waits for the process to finish executing
+     *
+     * This will in turn call wait(2) on the child process, awaiting for it
+     * to complete.  This function _must_ be called on all created
+     * `external_process` objects.  This can only be called once.
```

Comment:
What state is it in if the result is a failed future? Do you still have to unconditionally call `terminate`?

Reply:
Kind of depends.  If the future fails (because the wait(2) system call fails) then it's possible the child has already been reaped or it was never successfully started.  Which means wait() was called >1 which shouldn't happen.